torch/utils/data/distributed.py

```diff
@@ -65,3 +65,117 @@ def __len__(self):
 
     def set_epoch(self, epoch):
         self.epoch = epoch
+
+class ChunkDataReader(object):
+    r"""Reads a chunk of data given a chunk index.
+
+    A chunk could be full file, section of a large file, folders, URLs, or any other
+    abstraction that allows data to be segmented roughtly the same size.
+
+    As an example, chunking could be used in a scenario where MNIST dataset for training 
+    would be converted in CSV text files and equally splitted in several files,
+    e.g. 60 files with 1000 lines each.  In this case, each worker would read
+    a text file with all 1000 images.
+
+    Another example would be chunks as individual folders, containing multiple audio files.
+    In this scenario, each worker would have access to different folders, 
+    where binary files would be read from the filesystem.
+
+    A chunk doesn necessarily need to rely in multiple files to split data laoding across workers.
+    It is possible to extract chunks from a single large dataset file, where each worker would
+    seek to specific positions to load data.
+
+    The reading logic must be implemented inside `__call__` method
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self, idx):
+        r"""Returns `list(samples)`
+
+        If `samples` contain tensors, numpy arrays, numbers,
+        dicts or lists, the default collate function can be implicitly used.
+        Otherwise, a custom `collate_fn` must be provided to `DataLoader`
+
+        When no more data is available, `StopIteration` must be raised.
+        """
+        raise NotImplementedError
+
+class DistributedChunkSampler(Sampler):
+    r"""This sampler introduces distributed sampling without padding and dataset dependency.
+
+    Python DataLoader uses multi-processing instead of
+    multi-threading for parallelism, therefore, each worker
+    is a separate process with an identical copy of sampler.
+    Because of that, on a distributed environment, different processes
+    could read the same portion of the dataset. This sampler uses strides,
+    based on the global worker rank concept, to coordinate parallel reading of chunks of data.
+
+    Each instance of `DistributedChunkSampler` must be configured with a different
+    `global_worker_rank` (aka strides), so that chunk sampling happens in a round-robin
+    fashion on each worker, e.g. rank0, rank1, ..., rankN, rank0, rank1, ..., rankN, ...
+
+    For example, assume 2 workers reading the same `ChunkDataset` dataset.
```

Comment:
Please include an example of the worker_initialization function here. I would take a general case, two nodes each with two workers (total fo 4 workers)

Reply:
When I first wrote this, I though about this... worker initialization function is used to assign a global_worker_rank, only and it is dependent on specific user scenario. It is not strictly necessary to use dataloader workers > 1 to use chunk dataset, so adding worker initialization code here could give the wrong idea.

The paragraph at distributed.py line 130 was an attempt to replace the worker initialization snippet by a formula that explains how the global_worker_rank should be assigned, rather on a worker initialization function or main training script